src/com/ms/silverking/cloud/zookeeper/ZooKeeperExtended.java

```diff
@@ -42,1024 +43,1081 @@
+    private Stopwatch retrySW;
+    private Stopwatch timeoutSW;
+    private Set<String> errorKeys;
+    private Stopwatch displayMissingSW;
+
+    AsyncGet(Set<String> paths, Watcher watcher, CancelableObserver observer) {
+      this.paths = paths;
+      this.dataMap = new ConcurrentHashMap<>();
+      this.errorKeys = new ConcurrentSkipListSet<>();
+      this.lock = new ReentrantLock();
+      this.cv = lock.newCondition();
+      this.watcher = watcher;
+      this.observer = observer;
+      retrySW = new SimpleStopwatch();
+      timeoutSW = new SimpleStopwatch();
+      displayMissingSW = new SimpleStopwatch();
     }
-    
-    public Map<String,byte[]> getByteArrays(Set<String> paths, Watcher watcher) throws KeeperException {
-        AsyncGet    aGet;
-        
-        aGet = new AsyncGet(paths, watcher, null);
-        aGet.issueGets();
-        aGet.waitForCompletion();
-        return aGet.getDataMap();
+
+    AsyncGet(Set<String> paths) {
+      this(paths, null, null);
     }
-    
-    public Map<String,byte[]> getByteArrays(String basePath, Set<String> children, Watcher watcher, CancelableObserver observer) throws KeeperException {
-        ImmutableSet.Builder<String>    childrenPaths;
-        AsyncGet    aGet;
-        
-        childrenPaths = ImmutableSet.builder();
-        for (String child : children) {
-            childrenPaths.add(basePath +"/"+ child);
-        }
-        
-        aGet = new AsyncGet(childrenPaths.build(), watcher, observer);
-        aGet.issueGets();
-        aGet.waitForCompletion();
-        
-        return stripBasePathFromKeys(aGet.getDataMap(), basePath);
+
+    Map<String, byte[]> getDataMap() {
+      return dataMap;
     }
-    
-    public Map<String,byte[]> getByteArrays(String basePath, Set<String> children) throws KeeperException {
-        return getByteArrays(basePath, children, null, null);
+
+    void issueGets() {
+      for (String path : paths) {
+        //System.out.println("g: "+ path);
+        getData(path, watcher, this, null);
+      }
+      retrySW.reset();
     }
-    
-    private Map<String, byte[]> stripBasePathFromKeys(Map<String, byte[]> rawMap, String basePath) {
-        ImmutableMap.Builder<String, byte[]>    strippedMap;
-        
-        strippedMap = ImmutableMap.builder();
-        for (Map.Entry<String, byte[]> entry : rawMap.entrySet()) {
-            strippedMap.put(stripBasePath(entry.getKey(), basePath), entry.getValue());
+
+    public void issueGetsForIncomplete() {
+      if (observer == null || observer.isActive()) {
+        for (String path : paths) {
+          if (!dataMap.containsKey(path) && !errorKeys.contains(path)) {
+            Log.warning("Issuing sync zk get: ", path);
+            try {
+              byte[] data;
+
+              data = getData(path, watcher, null);
+              if (data != null) {
+                Log.warning("Sync zk get success: ", path);
+                dataMap.put(path, data);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **611 and 619**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).